src/maxtree.jl

```diff
@@ -0,0 +1,881 @@
+"""
+Max-tree morphological representation of an image.
+
+# Type parameters
+- `N`: the dimensions of the source image
+- `A`: the type of the `axes(image)`
+
+# Details
+Let's consider image *thresholding* operation. The result is the image mask s.t.
+``image[p] ≥ threshold`` for all ``p`` in the mask. This mask could also be
+viewed as a set of connected components. When *image thresholding* is
+sequentially applied for all possible thresholds, it generates a collection of
+connected components that could be organized into a hierarchical structure
+called *component tree*. A connected component at one threshold is a parent to
+a component at the higher threshold if the latter is a subset of the first.
+
+A *max-tree* is an efficient representation of the *component tree*.
+A connected component at one level is represented by the single *reference pixel*
+from this level, which is the parent to all other pixels of that level and
+to the *reference pixel* of the level above.
+
+The *max-tree* is the basis for many morphological operators,
+namely connected operators. Unlike morphological openings and closings, these
+operators do not require a fixed structuring element, but rather act with a
+flexible structuring element that meets a certain criterion.
+
+# See also
+[`area_opening`](@ref), [`area_closing`](@ref),
+[`diameter_opening`](@ref), [`diameter_closing`](@ref).
+
+# References
+1. Salembier, P., Oliveras, A., & Garrido, L. (1998). *Antiextensive Connected Operators for Image and Sequence Processing*. IEEE Transactions on Image Processing, 7(4), 555-570.
+   > https://doi.org/10.1109/83.663500
+2. Berger, C., Geraud, T., Levillain, R., Widynski, N., Baillard, A., Bertin, E. (2007). *Effective Component Tree Computation with Application to Pattern Recognition in Astronomical Imaging*. In International Conference on Image Processing (ICIP), 41-44.
+   > https://doi.org/10.1109/ICIP.2007.4379949
+3. Najman, L., & Couprie, M. (2006). *Building the component tree in quasi-linear time*. IEEE Transactions on Image Processing, 15(11), 3531-3539.
+   > https://doi.org/10.1109/TIP.2006.877518
+4. Carlinet, E., & Geraud, T. (2014). *A Comparative Review of Component Tree Computation Algorithms*. IEEE Transactions on Image Processing, 23(9), 3885-3895.
+   > https://doi.org/10.1109/TIP.2014.2336551
+"""
+struct MaxTree{N,A}
+    """
+    Axes of the source image.
+    """
+    axes::A
```

Comment:
The idea is to preserve the axes of the original image, which may use the indexing different from `parents::Array{T, N}`.
`.axes` is used by the performance-critical `isvalid_offset()`, so the types of the input arguments have to be inferrable from `rebuild!()` to allow inlining and efficient code generation of `isvalid_offset()`.
However, I don't have a real use case for non-standard axes, so this feature is half-baked at the moment.
Still, I think it doesn't introduce a lot of complexity, so we may keep it for the future.

Reply:
Supporting arrays with offsets is tricky here. Since `MaxTree` doesn't store the original image, any algorithm still needs to accept `tree` and `img` as arguments. Thus if `img` is changed after the construction of maxtree, it's not guaranteed that it works correctly and there're no codes to detect that. That's said, this design doesn't really solve the problem.

I see two options to prevent the issue of inconsistent axes:

* store the original image in `MaxTree`, it doesn't require additional memory allocation if I understand it correctly.
* make the construction lazily (I don't have a specific design for this but I think it's doable )

Reply:
I've removed `A` and `axes` field. In fact, `MaxTree` only stores linear indices of the original image, and morphological operations use linear indices only.
So, as long as the linear indices of the original `image`, of `maxtree.parentindices` and of `output`, are compatible (I assume it should be so when `size(image) == size(maxtree.parentindices) == size(output)`), `MaxTree` would give the correct results for arrays with non-standard offsets.

Reply:
> I assume it should be so when `size(image) == size(maxtree.parentindices) == size(output)`

Some power users (perhaps only you @alyst :P ) might figure out how to register the offsets, but for a more reliable and simpler usage, it should be `axes(image) == axes(maxtree.parentindices) == axes(output)`, since `parentindices` is defined as `Array{Int,N}`, I would say offset arrays are actually not supported.

My suggestion is to echo @timholy and to not support OffsetArray in this very first PR, remove related codes designed for offsets, raise errors for offset arrays,  and make the implementation simpler.

Reply:
Mostly I'd say that I don't want to drag anyone through implementing something difficult that isn't used by the author, so I propose those as a safety net for an implementation which has specific requirements. But if there's enthusiasm to support arbitrary indexing from the get-go, I'm all in favor.

The main thing I'd say is to leave arrays in the shape they have; yes you can use linear indexing but no you shouldn't shape things as vectors. That's because you lose an enormous amount of information when you store a multidimensional object with arbitrary indexing as a one-dimensional object.

Reply:
I agree that arrays of the same shape as the original image are more useful. I've updated `areas` and `diameters` accordingly.